ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsMaterializedWindowTableIQv2.java

```diff
@@ -0,0 +1,190 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams.materialization.ks;
+
+import static org.apache.kafka.streams.query.StateQueryRequest.inStore;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import com.google.common.collect.Range;
+import com.google.common.collect.Streams;
+import io.confluent.ksql.GenericKey;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.streams.materialization.MaterializationException;
+import io.confluent.ksql.execution.streams.materialization.MaterializedWindowedTable;
+import io.confluent.ksql.execution.streams.materialization.WindowedRow;
+import io.confluent.ksql.util.IteratorUtil;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.kstream.internals.TimeWindow;
+import org.apache.kafka.streams.query.StateQueryRequest;
+import org.apache.kafka.streams.query.StateQueryResult;
+import org.apache.kafka.streams.query.WindowKeyQuery;
+import org.apache.kafka.streams.query.WindowRangeQuery;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.apache.kafka.streams.state.WindowStoreIterator;
+
+/**
+ * Kafka Streams impl of {@link MaterializedWindowedTable}.
+ */
+class KsMaterializedWindowTableIQv2 implements MaterializedWindowedTable {
+
+  private final KsStateStore stateStore;
+  private final Duration windowSize;
+
+  KsMaterializedWindowTableIQv2(final KsStateStore store, final Duration windowSize) {
+    this.stateStore = Objects.requireNonNull(store, "store");
+    this.windowSize = Objects.requireNonNull(windowSize, "windowSize");
+  }
+
+  @Override
+  public List<WindowedRow> get(
+      final GenericKey key,
+      final int partition,
+      final Range<Instant> windowStartBounds,
+      final Range<Instant> windowEndBounds
+  ) {
+    try {
+      final Instant lower = calculateLowerBound(windowStartBounds, windowEndBounds);
+      final Instant upper = calculateUpperBound(windowStartBounds, windowEndBounds);
+      final WindowKeyQuery<GenericKey, ValueAndTimestamp<GenericRow>> query =
+          WindowKeyQuery.withKeyAndWindowStartRange(key, lower, upper);
+      final StateQueryRequest<WindowStoreIterator<ValueAndTimestamp<GenericRow>>> request =
+          inStore(stateStore.getStateStoreName()).withQuery(query);
+      final KafkaStreams streams = stateStore.getKafkaStreams();;
+      final StateQueryResult<WindowStoreIterator<ValueAndTimestamp<GenericRow>>> result =
+          streams.query(request);
+
+      try (WindowStoreIterator<ValueAndTimestamp<GenericRow>> it
+          = result.getGlobalResult().getResult()) {
```

Comment:
Here, you'll want to use the `partition` to get the specific partition result (as above, this isn't a global store, so there won't be a global result).

Reply:
Makes sense, done.